scalyr/scalyr-agent-2
# Improve rate limiter

# Background

Previous rate limiter adjustment was too simplistic.  In particular, a string of successes leads to rate increase to max.  If customer supplies a huge max number, upon server failure, it will take longer to backoff.

This PR installs multiple checks to rate limiter as follows:

1. Hardcoded min/max limits -- if user enters min/max beyond these limits, the rate limiter will adjust them and emit a warning.

2. reset-then-multiply strategy now applies both ways -- upon reversal, reset to initial rate, then increase/decrease thereafter

3. whenever performing rate adjustment, always take the more conservative value between actual rate and target rate, and then adjust by multiplication factor.   E.g. if current target rate is 10rps, but actual rate was only 6rps, come increase time, we multiply 2.0 (increase factor) by 6, not 10.  Furthermore, when increasing, the new target rate can never be lower than current target rate.
Same logic for reverse direction.

4. There is still asymmetry in that increases require a string of successes whereas backoff requires only 1 failure.

# How tested?

The original rate limiter unit tests pass. (I am adding new ones)
I am still working on the unit test but posting this PR so you can review the core logic.

I have also manually tested by faking 404s in the `query_api` function where I write to a text file on disk and the agent pretends that the API masters are temporarily unable.  In this way, visualize the rate adjustment real time.  Here is a example debug log of real-time rate adjustment (starting rate is 20K -- for context, the huge test cluster I ran could easily support a rate of 200K even though cpu was 65% baseline.)

# Manual testing

1. Here we simulate a user accidentally setting max to huge number. It is auto-restricted to 20K.
2. Then, initial rate is also auto-restricted to 20K.
3. The system easily handles 20K rps for a while.
4. I then `echo "404" > simfile` in the agent log dir. The system detects this and pretends it sees failures.
5. As you can see from the log, backoff occurs quickly.
6. Within seconds, I delete the simfile, pretending to restore health to api masters
7. The system then gradually becomes more aggressive.

Finally, note that when you see `156.25 (166.812098987) x 2.0 -> 312.5`, the number to left of the parens is the current target rate.  The number in parens is the actual calculated rate.  Since we're increasing, we use the conservative (smaller of the two).  The new target rate becomes 312.5.  Since it is lower than max, it's allowed.

![image](https://user-images.githubusercontent.com/48775713/59549233-6ac3a980-8f0f-11e9-8513-29bd1205ac29.png)
